File: common/src/main/java/dev/obscuria/fragmentum/common/signal/_Signal4.java
package dev.obscuria.fragmentum.common.signal;

import dev.obscuria.fragmentum.v2.api.common.signal.Signal4;

public final class _Signal4<P1, P2, P3, P4> extends _Signal<Signal4.Listener<P1, P2, P3, P4>> implements Signal4<P1, P2, P3, P4> {

    @Override public void emit(P1 p1, P2 p2, P3 p3, P4 p4) {
        emitInternal(it -> it.consume(p1, p2, p3, p4));
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/tools/signal/Signal4.kt
package dev.obscuria.fragmentum.tools.signal

class Signal4<P1, P2, P3, P4> : Signal<Signal4.Listener<P1, P2, P3, P4>>() {

    fun emit(p1: P1, p2: P2, p3: P3, p4: P4) = emit { it.consume(p1, p2, p3, p4) }

    fun interface Listener<P1, P2, P3, P4> {

        fun consume(p1: P1, p2: P2, p3: P3, p4: P4)
    }
}